crates/noirc_evaluator/src/ssa/function.rs

```diff
@@ -0,0 +1,208 @@
+use std::collections::HashMap;
+
+use crate::environment::Environment;
+use acvm::acir::OPCODE;
+use acvm::FieldElement;
+use noirc_frontend::hir_def::expr::HirCallExpression;
+use noirc_frontend::hir_def::function::Parameters;
+use noirc_frontend::hir_def::stmt::HirPattern;
+use noirc_frontend::node_interner::FuncId;
+
+use super::{
+    code_gen::IRGenerator,
+    context::SsaContext,
+    node::{self, NodeId, ObjectType},
+    ssa_form,
+};
+
+pub struct SSAFunction<'a> {
+    pub igen: IRGenerator<'a>,
+    pub id: FuncId,
+    //signature..
+    pub arguments: Vec<NodeId>,
+}
+
+impl<'a> SSAFunction<'a> {
+    //Parse the AST function body into ssa form in cfg
+    pub fn parse_statements(
+        &mut self,
+        block: &[noirc_frontend::node_interner::StmtId],
+        env: &mut Environment,
+    ) {
+        for stmt in block {
+            self.igen.evaluate_statement(env, stmt).unwrap();
+        }
+    }
+
+    pub fn new(func: FuncId, ctx: &'a noirc_frontend::hir::Context) -> SSAFunction<'a> {
+        SSAFunction {
+            igen: IRGenerator::new(ctx),
+            id: func,
+            arguments: Vec::new(),
+        }
+    }
+
+    pub fn compile(&mut self) -> Option<NodeId> {
+        //Optimisation
+        super::block::compute_dom(&mut self.igen.context);
+        super::optim::cse(&mut self.igen.context);
+        //Unrolling
+        super::flatten::unroll_tree(&mut self.igen.context);
+        super::optim::cse(&mut self.igen.context)
+    }
+
+    //generates an instruction for calling the function
+    pub fn call(
+        func: FuncId,
+        arguments: &[noirc_frontend::node_interner::ExprId],
+        eval: &mut IRGenerator,
```

Comment:
Why is this IRGenerator named eval rather than igen - is there any significance?

Reply:
No, I am not good with naming :)
I suppose the idea was because it evaluates the AST, but that was before the renaming into IRgenerator and it is better to use consistent naming so I renamed it to igen.